[GENFI] Update genfi patch

MatthieuJoulot opened the pull request:
Processing the whole dataset reveals a few more problems. This PR handles these problems.

MatthieuJoulot commented:
The test should break since the CI needs to be updated

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -795,10 +815,11 @@ def merge_philips_diffusion(
-        json.dump(data, json_file, indent=4)
+    with open(json_file, "r+") as f:
+        if multipart_id is not None:
+            data = json.load(f)
```
Why these changes ?

MatthieuJoulot replied:
I don't know why, but this doesn't work. So I changed back to what I had done in the first place, which I does !
It may not be optimal though, so we can look at it together

NicolasGensollen replied:
OK, we'll take a look at the error you're getting this afternoon.

NicolasGensollen commented on docs/Converters/GENFItoBIDS.md:
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
 
+## Downloading GENFI
+
+To download GENFI in a way that you can convert it, you only need to make sure of two thing: 
```
```suggestion
To download GENFI in a way that you can convert it, you need to make sure of two things: 
```

NicolasGensollen commented on docs/Converters/GENFItoBIDS.md:
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
+## Downloading GENFI
+
+To download GENFI in a way that you can convert it, you only need to make sure of two thing: 
+- To only select the "simplify downloaded archive structure" in download data options.
```
```suggestion
- Only select the "simplify downloaded archive structure" in download data options.
```

MatthieuJoulot commented:
Unit-tests currently crash for the following reason: I changed the content of the `bidsignore` and added `dwi` to be ignored. I did that in order to respect the Philips sequence and to pass the bids-validator, which otherwise contradict.

They contradict because GENFI acquired a single b0, which has no bvecs or bvals, which is not allowed by the bids-spec.

I'm conflicted regarding this addition of dwi to the bidsignore, I'd rather remove it (even though I've added it), and not let GENFI pass the bids-validator, since it is out of spec.
@NicolasGensollen WDYT ?

NicolasGensollen commented on docs/Converters/GENFItoBIDS.md:
```diff
@@ -17,6 +17,14 @@ Please note that this converter currently processes the following modalities :
+
+To download GENFI in a way that you can convert it, you only need to make sure of two thing: 
+- To only select the "simplify downloaded archive structure" in download data options.
+- And in Select Image Data section Scans Format, select only "DICOM"
```
I'm not sure my suggestion is correct, please verify where the `"` should be placed.

```suggestion
- Select only "DICOM" in "Select Image Data" section "Scans Format".
```

NicolasGensollen reviewed (commented):
Thanks @MatthieuJoulot !
I made a quick first pass without testing it (will do that later today...).
I'm a bit concerned with the changes made to utility functions used across all converters.
Did you make sure your modifications were not breaking anything ?

MatthieuJoulot commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -640,15 +648,23 @@ def merge_imaging_data(df_dicom: DataFrame) -> DataFrame:
+        # ].agg("/".join, axis=1),
+    )
+    df_sub_ses_run = df_sub_ses_run.assign(
+        bids_filename=lambda df: df.bids_filename.apply(lambda x: x.replace("__", "_"))
```
because of the task_rest that is aggregated as well. since it only exists for fMRI acquisitions, two underscores are aggregated for other modalities.

NicolasGensollen commented on clinica/iotools/bids_utils.py:
```diff
@@ -905,6 +911,9 @@ def run_dcm2niix(
         )
+        return False
+    else:
+        return True
```
```suggestion
    return True
```

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -709,13 +721,17 @@ def write_bids(
-        if "dwi" in metadata["bids_filename"] and metadata.manufacturer == "Philips":
+        if (
+            "dwi" in metadata["bids_filename"]
+            and "Philips" in metadata.manufacturer
```
Why the change from equality to a weaker inclusion here ?

MatthieuJoulot replied:
Sometimes the manufacturer is not just `Philips`but `Philips &Co`or whatnot

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -835,7 +852,7 @@ def _get_multipart_id(nb_parts: PhilipsNumberOfParts, run_num: str) -> Optional[
         raise ValueError(f"{run_num} is outside of the scope.")
     if nb_parts == PhilipsNumberOfParts.FIVE:
-        if run_num in (f"run-0{k}" for k in range(1, 5)):
+        if run_num in (f"run-0{k}" for k in range(1, 6)):
```
Was the 5 a mistake or has something changed ?

MatthieuJoulot replied:
No ! We want to have the 5 first runs

NicolasGensollen commented on clinica/iotools/bids_utils.py:
```diff
@@ -885,6 +886,11 @@ def run_dcm2niix(
+
+    Returns
+    -------
+    dcm2niix_success: bool
```
Since you're not naming the return value:

```suggestion
    bool :
```

NicolasGensollen reviewed (commented):
Made a second pass. I think we need to refactor a bit the function `merge_imaging_data`. It's getting really difficult to understand what's going on in it and I'm afraid it will look totally opaque to us in a few months...

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
+def _compute_source_id_and_source_ses_id(df: DataFrame) -> DataFrame:
+    """Adds two columns built from the column 'source'.
+
+    - 'source_id': subject ID in the raw dataset
```
Could you add a line for `source` as well ?

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
+
+    Example
+    -------
+    source_id = 'C9ORF001'
```
Same, it'd be great to have the `source` column in your example

NicolasGensollen commented on clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py:
```diff
@@ -257,13 +262,85 @@ def read_imaging_data(source_path: PathLike) -> DataFrame:
+    Example
+    -------
+    source_id = 'C9ORF001'
+    source_ses_id = '11
```
Why 11 ?

NicolasGensollen replied:
It's much clearer now that you added an example for `source`, thanks !

NicolasGensollen reviewed (approved):
LGTM, thanks @MatthieuJoulot !

NicolasGensollen commented:
Alright ! All tests are passing, merging...
